datalad/distribution/dataset.py

```diff
@@ -236,16 +239,19 @@ def id(self):
         for different versions of the same dataset (that have all been derived
         from the same original dataset repository).
 
+        Note, that a plain git/git-annex repository doesn't necessarily have
+        a dataset id yet. It is create by `Dataset.create()` and stored in
+        .datalad/config. If None is returned while there is a valid repository,
+        there may have never been a call to `create` in the history of commit
+        HEAD.
+
         Returns
         -------
         str
           This is either a stored UUID, or `None`.
         """
-        if self._id is None:
-            # if we have one on record, stick to it!
-            self.config.reload()
-            self._id = self.config.get('datalad.dataset.id', None)
-        return self._id
+
+        return self.config.get('datalad.dataset.id', None)
```

Comment:
I don't see a reason to reload when we access `Dataset.id`. The important part is to ask the correct `ConfigManager` instance instead of assuming that it never changes. In fact the `reload()` before was quite pointless, since the only time it was called was the first access to `id` which is by no means a significant or even valid point in time regarding the properties of the dataset. Try this on master:

```
$> python -c 'from datalad.api import Dataset; ds = Dataset("/tmp/notexistent"); print(ds.id)'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/ben/Development/datalad/datalad/distribution/dataset.py", line 246, in id
    self.config.reload()
  File "/home/ben/Development/datalad/datalad/distribution/dataset.py", line 261, in config
    self._cfg = ConfigManager(dataset=self, dataset_only=False)
  File "/home/ben/Development/datalad/datalad/config.py", line 204, in __init__
    self.reload(force=True)
  File "/home/ben/Development/datalad/datalad/config.py", line 258, in reload
    stdout, stderr = self._run(run_args, log_stderr=True)
  File "/home/ben/Development/datalad/datalad/config.py", line 526, in _run
    out = self._runner.run(['git', 'config'] + args, **kwargs)
  File "/home/ben/Development/datalad/datalad/cmd.py", line 673, in run
    cmd, env=self.get_git_environ_adjusted(env), *args, **kwargs)
  File "/home/ben/Development/datalad/datalad/cmd.py", line 487, in run
    stdin=stdin)
  File "/usr/lib/python3.5/subprocess.py", line 676, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.5/subprocess.py", line 1282, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/notexistent'
```
And even if we would fix for not crashing it would need to report `None` as its id. Then call `ds.create()` and `ds.id` again with old implementation of `id`. It would still be `None`, which is wrong. See also: #2948

Reply:
Also note, that `self.config.get('datalad.dataset.id', None)` does NOT mean to query from FS. `ConfigManager` caches. This might look more expensive than it is. The actual cost comes indirectly via `repo` (triggered by `config`).